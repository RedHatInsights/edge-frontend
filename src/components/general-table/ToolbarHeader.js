import React from 'react';
import {
  Toolbar,
  Pagination,
  ToolbarItem,
  ToolbarContent,
  Button,
  ToggleGroup,
  ToggleGroupItem,
  Skeleton,
} from '@patternfly/react-core';
import PropTypes from 'prop-types';
import FilterControls from './FilterControls';
import FilterChip from './FilterChips';
import ToolbarKebab from './ToolbarKebab';

const ToolbarButtons = ({ buttons }) => {
  return buttons.map(({ title, click, isDisabled, id }, index) => (
    <ToolbarItem key={index}>
      <Button onClick={click} variant="primary" isDisabled={isDisabled} id={id}>
        {title}
      </Button>
    </ToolbarItem>
  ));
};

const ToolbarHeader = ({
  toolbarButtons,
  filters,
  setFilterValues,
  filterValues,
  chipsArray,
  setChipsArray,
  isLoading,
  count,
  perPage,
  setPerPage,
  page,
  setPage,
  toggleButton,
  toggleAction,
  toggleState,
  children,
  kebabItems,
}) => {
  return (
    <Toolbar id="toolbar-header" data-testid="toolbar-header-testid">
      <ToolbarContent>
        <FilterControls
          filters={filters}
          filterValues={filterValues}
          setFilterValues={setFilterValues}
        >
          {children}
        </FilterControls>
        {toolbarButtons && <ToolbarButtons buttons={toolbarButtons} />}
        {toggleButton && (
          <ToggleGroup>
            {toggleButton.map((btn) => (
              <ToggleGroupItem
                key={btn.key}
                text={btn.title}
                isSelected={toggleState === btn.key}
                onChange={() => toggleAction(btn.key)}
              />
            ))}
          </ToggleGroup>
        )}
        {kebabItems && <ToolbarKebab kebabItems={kebabItems} />}
        <ToolbarItem variant="pagination" align={{ default: 'alignRight' }}>
          {isLoading ? (
            <Skeleton width="200px" />
          ) : count > 0 ? (
            <Pagination
              data-testid="pagination-header-test-id"
              itemCount={count}
              perPage={perPage}
              page={page}
              onSetPage={(_e, pageNumber) => setPage(pageNumber)}
              widgetId="pagination-options-menu-top"
              onPerPageSelect={(_e, perPage) => setPerPage(perPage)}
              isCompact
            />
          ) : null}
        </ToolbarItem>
      </ToolbarContent>
      <ToolbarContent>
        <ToolbarItem variant="chip-group" spacer={{ default: 'spacerNone' }}>
          <FilterChip
            filterValues={filterValues}
            setFilterValues={setFilterValues}
            chipsArray={chipsArray}
            setChipsArray={setChipsArray}
            setPage={setPage}
          />
        </ToolbarItem>
      </ToolbarContent>
    </Toolbar>
  );
};

ToolbarHeader.propTypes = {
  filters: PropTypes.array,
  toolbarButtons: PropTypes.array,
  setFilterValues: PropTypes.func,
  filterValues: PropTypes.array,
  input: PropTypes.string,
  count: PropTypes.number,
  perPage: PropTypes.number,
  setPerPage: PropTypes.func,
  page: PropTypes.number,
  setPage: PropTypes.func,
  chipsArray: PropTypes.array,
  setChipsArray: PropTypes.func,
  isLoading: PropTypes.bool,
  toggleButton: PropTypes.array,
  toggleAction: PropTypes.func,
  toggleState: PropTypes.number,
  children: PropTypes.element,
  kebabItems: PropTypes.array,
};
export default ToolbarHeader;
