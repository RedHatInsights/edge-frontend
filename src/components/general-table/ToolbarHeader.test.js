import React from 'react';
import ToolbarHeader from './ToolbarHeader';
import { render } from '@testing-library/react';

describe('Toolbar Header', () => {
  it('should render correctly', async () => {
    const toolbarButtons = [
      {
        title: 'Test toolbar button',
        click: () => jest.fn(),
      },
    ];
    const filters = [
      {
        label: 'Test 1',
        type: 'text',
      },
    ];
    const filterValues = [
      { type: 'text', label: 'Test 1', value: 'test-label' },
    ];
    const chipsArray = [{ key: 'Test 1', label: 'test-label' }];
    const setFilterValues = jest.fn();
    const setChipsArray = jest.fn();
    const setPage = jest.fn();
    const setPerPage = jest.fn();

    const { container, findByTestId, findByPlaceholderText, findByRole } =
      render(
        <ToolbarHeader
          count={100}
          toolbarButtons={toolbarButtons}
          filters={filters}
          filterValues={filterValues}
          setFilterValues={setFilterValues}
          chipsArray={chipsArray}
          setChipsArray={setChipsArray}
          isLoading={false}
          perPage={50}
          setPerPage={setPerPage}
          page={1}
          setPage={setPage}
          checkBoxState={{ hasCheckbox: false }}
          setCheckBoxState={() => jest.fn()}
        />
      );

    const headerElement = await findByTestId('toolbar-header-testid');
    const headerInput = await findByPlaceholderText('Filter by test 1');
    const headerButton = await findByRole('button', {
      name: 'Test toolbar button',
    });
    const paginationElement = await findByTestId('pagination-header-test-id');

    expect(headerElement).toBeDefined();
    expect(headerInput.value).toEqual('test-label');
    expect(headerButton).toBeDefined();
    expect(paginationElement.children[0].innerHTML).toEqual(
      '<b>1 - 50</b> of <b>100</b> '
    );

    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
