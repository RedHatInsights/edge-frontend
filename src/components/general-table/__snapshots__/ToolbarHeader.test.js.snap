// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Header should render correctly 1`] = `
<div
  class="pf-v5-c-toolbar"
  data-ouia-component-id="OUIA-Generated-Toolbar-1"
  data-ouia-component-type="PF5/Toolbar"
  data-ouia-safe="true"
  data-testid="toolbar-header-testid"
  id="toolbar-header"
  style="padding: 0px;"
>
  <div
    class="pf-v5-c-toolbar__content"
  >
    <div
      class="pf-v5-c-toolbar__content-section"
    >
      <div
        class="pf-v5-c-toolbar__item"
        data-testid="filter-input-testid"
      >
        <div
          class="pf-v5-c-input-group"
        >
          <div
            class="pf-v5-c-input-group__item"
          >
            <div
              class="pf-v5-c-text-input-group"
              id="textInput1"
              type="search"
            >
              <div
                class="pf-v5-c-text-input-group__main pf-m-icon"
              >
                <span
                  class="pf-v5-c-text-input-group__text"
                >
                  <span
                    class="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Select input for test 1"
                    class="pf-v5-c-text-input-group__text-input"
                    name="textInput1"
                    placeholder="Filter by test 1"
                    type="text"
                    value="test-label"
                  />
                </span>
              </div>
              <div
                class="pf-v5-c-text-input-group__utilities"
              >
                <button
                  aria-disabled="false"
                  aria-label="Reset"
                  class="pf-v5-c-button pf-m-plain"
                  data-ouia-component-id="OUIA-Generated-Button-plain-1"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 352 512"
                    width="1em"
                  >
                    <path
                      d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-c-toolbar__item"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary"
          data-ouia-component-id="OUIA-Generated-Button-primary-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Test toolbar button
        </button>
      </div>
      <div
        class="pf-v5-c-toolbar__item pf-m-pagination pf-m-align-right"
      >
        <div
          class="pf-v5-c-pagination pf-m-compact"
          data-ouia-component-id="OUIA-Generated-Pagination-top-1"
          data-ouia-component-type="PF5/Pagination"
          data-ouia-safe="true"
          data-testid="pagination-header-test-id"
          id="pagination-options-menu-top-top-pagination"
          style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
        >
          <div
            class="pf-v5-c-pagination__total-items"
          >
            <b>
              1
               - 
              50
            </b>
             
            of
             
            <b>
              100
            </b>
             
          </div>
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
              id="pagination-options-menu-top-top-toggle"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__text"
              >
                <b>
                  1
                   - 
                  50
                </b>
                 
                of
                 
                <b>
                  100
                </b>
                 
              </span>
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
          <nav
            aria-label="Pagination"
            class="pf-v5-c-pagination__nav"
          >
            <div
              class="pf-v5-c-pagination__nav-control"
            >
              <button
                aria-disabled="true"
                aria-label="Go to previous page"
                class="pf-v5-c-button pf-m-plain pf-m-disabled"
                data-action="previous"
                data-ouia-component-id="OUIA-Generated-Button-plain-2"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                disabled=""
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="pf-v5-c-pagination__nav-control"
            >
              <button
                aria-disabled="false"
                aria-label="Go to next page"
                class="pf-v5-c-button pf-m-plain"
                data-action="next"
                data-ouia-component-id="OUIA-Generated-Button-plain-3"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </button>
            </div>
          </nav>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pf-v5-c-toolbar__content"
  >
    <div
      class="pf-v5-c-toolbar__content-section"
    >
      <div
        class="pf-v5-c-toolbar__item pf-m-chip-group pf-m-spacer-none"
      >
        <div
          aria-labelledby="pf-random-id-0"
          class="pf-v5-c-chip-group pf-u-mr-xs pf-m-category"
          data-ouia-component-type="PF5/ChipGroup"
          data-ouia-safe="true"
          role="group"
        >
          <div
            class="pf-v5-c-chip-group__main"
          >
            <span
              class="pf-v5-c-chip-group__label"
              id="pf-random-id-0"
            >
              Test 1
            </span>
            <ul
              aria-labelledby="pf-random-id-0"
              class="pf-v5-c-chip-group__list"
              data-testid="filter-chip"
              role="list"
            >
              <li
                class="pf-v5-c-chip-group__list-item"
              >
                <div
                  class="pf-v5-c-chip"
                  data-ouia-component-id="OUIA-Generated-Chip-1"
                  data-ouia-component-type="PF5/Chip"
                  data-ouia-safe="true"
                >
                  <span
                    class="pf-v5-c-chip__content"
                  >
                    <span
                      class="pf-v5-c-chip__text"
                      id="pf-random-id-1"
                    >
                      test-label
                    </span>
                  </span>
                  <span
                    class="pf-v5-c-chip__actions"
                  >
                    <button
                      aria-disabled="false"
                      aria-label="close"
                      aria-labelledby="remove_pf-random-id-1 pf-random-id-1"
                      class="pf-v5-c-button pf-m-plain"
                      data-ouia-component-id="close"
                      data-ouia-component-type="PF5/Button"
                      data-ouia-safe="true"
                      id="remove_pf-random-id-1"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="pf-v5-svg"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 352 512"
                        width="1em"
                      >
                        <path
                          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                        />
                      </svg>
                    </button>
                  </span>
                </div>
              </li>
            </ul>
          </div>
        </div>
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-link"
          data-ouia-component-id="OUIA-Generated-Button-link-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Reset filters
        </button>
      </div>
    </div>
  </div>
  <div
    class="pf-v5-c-toolbar__content pf-m-hidden"
    hidden=""
  >
    <div
      class="pf-v5-c-toolbar__group"
    />
  </div>
</div>
`;
